python/tvm/meta_schedule/testing/distributed_measure_candidates.py

```diff
@@ -0,0 +1,199 @@
+        type=str,
+        default="nvidia/nvidia-v100",
+        help="Please specify the target hardware for tuning context.",
+    )
+    parser.add_argument(
+        "--rpc_host", type=str, help="Please provide the private IPv4 address for the tracker."
+    )
+    parser.add_argument(
+        "--rpc_port", type=int, default=4445, help="Please provide the port for the tracker."
+    )
+    parser.add_argument(
+        "--rpc_key",
+        type=str,
+        default="p3.2xlarge",
+        help="Please provide the key for the rpc servers.",
+    )
+    parser.add_argument(
+        "--builder_timeout_sec",
+        type=int,
+        default=10,
+        help="The time for the builder session to time out.",
+    )
+    parser.add_argument(
+        "--min_repeat_ms", type=int, default=100, help="The time for preheating the gpu."
+    )
+    parser.add_argument(
+        "--runner_timeout_sec",
+        type=int,
+        default=100,
+        help="The time for the runner session to time out.",
+    )
+    parser.add_argument(
+        "--batch_size",
+        type=int,
+        default=128,
+        help="The batch size of candidates sent to builder and runner each time.",
+    )
+    return parser.parse_args()
+
+
+# pylint: disable=too-many-locals
+def measure_candidates(database, builder, runner):
+    """Send the candidates to builder and runner for distributed measurement,
+    and save the results in a new json database.
+
+    Parameters
+    ----------
+    database : JSONDatabase
+        The database for candidates to be measured.
+    builder : Builder
+        The builder for building the candidates.
+    runner : Runner
+        The runner for measuring the candidates.
+
+    Returns
+    -------
+    None
+    """
+    candidates, runner_results, build_fail_indices, run_fail_indices = [], [], [], []
+    build_time, run_time = 0.0, 0.0
+    context = ms.TuneContext(target=Target(args.target))
+    tuning_records = database.get_all_tuning_records()
+    for record in tuning_records:
+        candidates.append(context.get_measure_candidate(record.workload.mod, record.trace))
+    for idx in range(0, len(candidates), args.batch_size):
+        batch_candidates = candidates[idx : idx + args.batch_size]
+        context.set_measure_candidates(batch_candidates)
+        build_start_time = time.time()
+        context.send_to_builder(builder)
+        build_end_time = time.time()
+        context.send_to_runner(runner)
+        batch_runner_results = context.join()
+        run_end_time = time.time()
+        runner_results.extend(batch_runner_results)
+        for i, result in enumerate(context.builder_results):
+            if result.error_msg is None:
+                ms.utils.remove_build_dir(result.artifact_path)
+            else:
+                build_fail_indices.append(i + idx)
+        context.clear_measure_state()
+        build_time += build_end_time - build_start_time
+        run_time += run_end_time - build_end_time
```

Comment:
There is a cool thing called `ms.Profiler` which you might be interested in! In our usecase, it could be as easy as:

```python
with ms.Profiler() as profiler:
    for each batch:
        with ms.Profiler.timeit("build"):
            ...
        ...
        with ms.Profiler.timeit("run"):
            ...
for seg_name, seg_total_time in profiler.get():
    print(...)
```

Reply:
Thanks! addressed :-)